main.js

```diff
@@ -0,0 +1,910 @@
+                })
+            })
+        ])
+        checkTorrent.folder = folderPath
+        checkTorrent.title = title
+        checkTorrent.side = true
+        return checkTorrent
+    }
+
+    // we must use this function to seed a BEP46 torrent that we have already published before
+    async currentAddress(address){
+        let haveTorrent = this.findTheAddress(address)
+        if(haveTorrent){
+            return haveTorrent
+        }
+        let folderPath = this._internal + path.sep + address
+        if(!await fs.pathExists(folderPath)){
+            throw new Error('folder does not exist')
+        }
+        let checkTorrent = await Promise.race([
+            new Promise((resolve, reject) => {
+                setTimeout(() => {reject(new Error(address + ' took too long, it timed out'))}, this._status.timeout)
+            }),
+            new Promise((resolve, reject) => {
+                this.webtorrent.seed(folderPath, {destroyStoreOnDestroy: true}, torrent => {
+                    resolve(torrent)
+                })
+            })
+        ])
+        let checkProperty = await Promise.race([
+            new Promise((resolve, reject) => {
+                setTimeout(() => {
+                    this.webtorrent.remove(checkTorrent.infoHash, {destroyStore: false})
+                    reject(new Error(address + ' property took too long, it timed out, please try again with only the keypair without the folder'))
+                },this._status.timeout)
+            }),
+            new Promise((resolve, reject) => {
+                this.ownData(address, checkTorrent.infoHash).then(res => {
+                    resolve(res)
+                }).catch(error => {
+                    console.log(error)
+                    // most likely the infohash of this torrent does not match what we have, stop this torrent and reject the promise
+                    this.webtorrent.remove(checkTorrent.infoHash, {destroyStore: false})
+                    reject(error)
+                })
+            })
+        ])
+        delete checkProperty.infoHash
+        checkProperty.folder = folderPath
+        checkProperty.side = true
+        for(const prop in checkProperty){
+            checkTorrent[prop] = checkProperty[prop]
+        }
+        return checkTorrent
+    }
+
+    // download a regular non-BEP46 non-user created torrent by entering a 40 character infohash
+    async loadHash(hash){
+        let haveTorrent = this.findTheHash(hash)
+        if(haveTorrent){
+            return haveTorrent
+        }
+        // if user had torrent before, then empty the folder so previous data does not conflict with the infohash
+        let folderPath = this._external + path.sep + hash
+        // try {
+        //     await fs.emptyDir(folderPath)
+        // } catch (error) {
+        //     console.log(error)
+        // }
+        checkTorrent = await Promise.race([
+            new Promise((resolve, reject) => {
```

Comment:
i will get to this later, need some elaboration.

Reply:
Mainly, there's a lot of duplicate code for "reject with an error after a certain timeout", the function I proposed could clean up a lot of it and make the code more readable.

e.g. this code could look like

```JavaScript
checkTorrent = await Promise.race([
      rejectAfter(this.timeout, new Error(hash + ' took too long, it timed out'))
      new Promise((resolve, reject) => {
        this.webtorrent.add(hash, { path: folderPath, destroyStoreOnDestroy: true }, torrent => {
          resolve(torrent)
        })
      })
    ])
```

That said, I'm seeing `webtorrent.add()` invoke d a bunch of times with promises wrappers, so it might be good to add a method for that too.

e.g.

```JavaScript
async addToWebTorrent(...args) {
    return new Promise((resolve) => {
    this.webtorrent.add(...args, resolve)
     })
}
```

which could make the final code look like this:

```JavaScript
checkTorrent = await Promise.race([
      rejectAfter(this.timeout, new Error(hash + ' took too long, it timed out'))
      this.addToWebTorrent(hash, { path: folderPath, destroyStoreOnDestroy: true })
    ])
```

Which I think will be a lot easier to read and maintain going forward. 😁